File: ED del diego/Haskell/Krustal/DataStructures/Graph/DictionaryWeightedGraph.hs
----------------------------------------------
-- Estructuras de Datos.  2018/19
-- 2º Curso del Grado en Ingeniería [Informática | del Software | de Computadores].
-- Escuela Técnica Superior de Ingeniería en Informática. UMA
--
-- Examen 4 de febrero de 2019
--
-- ALUMNO/NAME:
-- GRADO/STUDIES:
-- NÚM. MÁQUINA/MACHINE NUMBER:
--
-- Weighted Graph implemented by using a dictionary from
-- sources to another dictionary from destinations to weights
----------------------------------------------

module DataStructures.Graph.DictionaryWeightedGraph
  ( WeightedGraph
  , WeightedEdge(WE)
  , empty
  , isEmpty
  , mkWeightedGraphEdges
  , addVertex
  , addEdge
  , vertices
  , numVertices
  , edges
  , numEdges
  , successors
  ) where

import Data.List(nub, intercalate)

import qualified DataStructures.Dictionary.AVLDictionary as D

data WeightedEdge a w  = WE a w a deriving Show

instance (Eq a, Eq w) => Eq (WeightedEdge a w) where
  WE u w v == WE u' w' v' = (u==u' && v==v' || u==v' && v==u')
                              && w == w'

instance (Eq a, Ord w) => Ord (WeightedEdge a w) where
  compare (WE _ w _) (WE _ w' _) = compare w w'

data WeightedGraph a w  = WG (D.Dictionary a (D.Dictionary a w))

-- fromJust
fromJust :: Maybe (D.Dictionary a w) -> D.Dictionary a w
fromJust Nothing = D.empty
fromJust (Just dic) = dic

empty :: WeightedGraph a w
empty = WG D.empty

addVertex :: (Ord a) => WeightedGraph a w -> a -> WeightedGraph a w
addVertex (WG dict) v = WG (D.insert v (D.empty) dict) -- Cuando añado un vértice solamente debo añadirlo al diccionario asociandole un diccionario vacío

addEdge :: (Ord a, Show a) => WeightedGraph a w -> a -> a -> w -> WeightedGraph a w
{-
addEdge (WG dic) v u w
  | not (D.isDefinedAt v dic) || not (D.isDefinedAt u dic) = error "El vértice no esta en el grafo"
  | otherwise = WG (D.insert v ws dic)
    where
      ws = D.insert u w (fromJust (D.valueOf v dic))
-}
addEdge (WG dic) o d p = aux (WG dic) o d p (vertices (WG dic))
    where
      aux (WG dic) o d p verts
        | (elem o verts) && (elem d verts) = (WG (D.insert o da dic))
        | otherwise = error "Falta algún vértice"
        where
          da = D.insert d p (fromJust (D.valueOf o dic))

edges :: (Eq a, Eq w) => WeightedGraph a w -> [WeightedEdge a w]
edges (WG dic) = edgesAc (D.keysValues dic) [] --funcion auxiliar que recibe todos los vertices y una lista vacia
  where
    edgesAc [] we = nub we --nub borra de una lista los elementos duplicados
    edgesAc (x:xs) we
      | D.isEmpty (snd x) = edgesAc xs we
      | otherwise = edgesAc xs (D.foldKeysValues (\u w res -> (WE v w u):res) we (snd x))
        where v = fst x

successors :: (Ord a, Show a) => WeightedGraph a w -> a -> [(a,w)]
successors (WG dic) v 
  | not (D.isDefinedAt v dic) = error "El vértice no está en el grafo"
  | otherwise = D.keysValues (fromJust (D.valueOf v dic)) 


-- NO EDITAR A PARTIR DE AQUÍ    
-- DON'T EDIT ANYTHING BELOW THIS COMMENT

vertices :: WeightedGraph a w -> [a]
vertices (WG d) = D.keys d

isEmpty :: WeightedGraph a w -> Bool
isEmpty (WG d) = D.isEmpty d

mkWeightedGraphEdges :: (Ord a, Show a) => [a] -> [WeightedEdge a w] -> WeightedGraph a w
mkWeightedGraphEdges vs es = wg'
  where
    wg = foldl addVertex empty vs
    wg' = foldr (\(WE u w v) wg -> addEdge wg u v w) wg es

numVertices :: WeightedGraph a w -> Int
numVertices = length . vertices

numEdges :: (Eq a, Eq w) => WeightedGraph a w -> Int
numEdges = length . edges

instance (Eq a, Show a, Eq w, Show w) => Show (WeightedGraph a w) where
  show wg  = "DictionaryWeightedGraph("++vs++", "++as++")"
   where
    vs  = "("++ intercalate ", " (map show (vertices wg)) ++")"
    as  = "(" ++ intercalate ", " (map showEdge (edges wg)) ++ ")"
    showEdge (WE x w y)  = intercalate "-" [ show x, show w, show y ]

File: ED del Pedro/ED/Java/Febrero 2019/src/dataStructures/graph/DictionaryWeightedGraph.java
/**----------------------------------------------
 * -- Estructuras de Datos.  2018/19
 * -- 2º Curso del Grado en Ingeniería [Informática | del Software | de Computadores].
 * -- Escuela Técnica Superior de Ingeniería en Informática. UMA
 * --
 * -- Examen 4 de febrero de 2019
 * --
 * -- ALUMNO/NAME:
 * -- GRADO/STUDIES:
 * -- NÚM. MÁQUINA/MACHINE NUMBER:
 * --
 * ----------------------------------------------
 */

package dataStructures.graph;

import java.util.Iterator;

import dataStructures.dictionary.Dictionary;
import dataStructures.dictionary.HashDictionary;

import dataStructures.set.Set;
import dataStructures.set.HashSet;
import dataStructures.tuple.Tuple2;

public class DictionaryWeightedGraph<V, W extends Comparable<? super W>> implements WeightedGraph<V, W> {

    static class WE<V1, W1 extends Comparable<? super W1>> implements WeightedEdge<V1, W1> {

		V1 src, dst;
        W1 wght;

        WE(V1 s, V1 d, W1 w) {
            src = s;
            dst = d;
            wght = w;
        }

        public V1 source() {
            return src;
        }

        public V1 destination() {
            return dst;
        }

        public W1 weight() {
            return wght;
        }

        public String toString() {
            return "WE(" + src + "," + dst + "," + wght + ")";
        }

		public int hashCode() {
            return src.hashCode() + dst.hashCode() + wght.hashCode();
		}

		public boolean equals(Object obj) {
            boolean ok = false;
            if (obj instanceof WE){
                WE other = (WE) obj;
                ok = this.src.equals(other.src) && this.dst.equals(other.dst) && this.wght.equals(other.wght);
            }
            return ok;
		}
		
		public int compareTo(WeightedEdge<V1, W1> o) {
			return wght.compareTo(o.weight());
		}
    }

    /**
     * Each vertex is associated to a dictionary containing associations
     * from each successor to its weight
     */
    protected Dictionary<V, Dictionary<V, W>> graph;

    public DictionaryWeightedGraph() {
        graph = new HashDictionary<>();
    }


    public void addVertex(V v) {
        // Añado el vértice con un diccionario vacío
        graph.insert(v,new HashDictionary<>());
    }

    public void addEdge(V src, V dst, W w) {
        // Ahora tengo que mirar que ambos vértices están en el grafo
        if(!graph.isDefinedAt(src) || !graph.isDefinedAt(dst)){
            throw new GraphException("Los vértices no están en el grafo");
        }
        // Añado en su diccionario asociado
        graph.valueOf(src).insert(dst,w);
    }

    public Set<Tuple2<V, W>> successors(V v) {
        if(!graph.isDefinedAt(v)){
            throw new GraphException("Este vértice no está en el grafo");
        }
        Set<Tuple2<V,W>> suc = new HashSet<>();
        for(Tuple2<V,W> kv: graph.valueOf(v).keysValues()){
            suc.insert(kv);
        }
        return suc;
    }


    public Set<WeightedEdge<V, W>> edges() {
        Set<WeightedEdge<V,W>> e = new HashSet<>();
        for(Tuple2<V,Dictionary<V,W>> elm: graph.keysValues()){
            for(Tuple2<V,W> aux: elm._2().keysValues()){
                e.insert(new WE<>(elm._1(), aux._1(), aux._2()));
            }
        }
        return e;
    }






    /** DON'T EDIT ANYTHING BELOW THIS COMMENT **/


    public Set<V> vertices() {
        Set<V> vs = new HashSet<>();
        for (V v : graph.keys())
            vs.insert(v);
        return vs;
    }


    public boolean isEmpty() {
        return graph.isEmpty();
    }

    public int numVertices() {
        return graph.size();
    }


    public int numEdges() {
        int num = 0;
        for (Dictionary<V, W> d : graph.values())
            num += d.size();
        return num / 2;
    }


    public String toString() {
        String className = getClass().getSimpleName();
        String s = className + "(vertices=(";

        Iterator<V> it1 = vertices().iterator();
        while (it1.hasNext())
            s += it1.next() + (it1.hasNext() ? ", " : "");
        s += ")";

        s += ", edges=(";
        Iterator<WeightedEdge<V, W>> it2 = edges().iterator();
        while (it2.hasNext())
            s += it2.next() + (it2.hasNext() ? ", " : "");
        s += "))";

        return s;
    }
}